machine.go

```diff
@@ -405,46 +448,43 @@ func (m *Machine) writerKernelModules(w *writerhelper.WriterHelper, moddir strin
 			"modules.symbols",
 			"modules.symbols.bin",
 			"modules.builtin.bin",
-			"modules.devname")
-
-	// build a list of built-in modules so that we don’t attempt to copy them
-	var builtinModules = make(map[string]bool)
+			"modules.devname"}
 
-	f, err := os.Open(path.Join(moddir, "modules.builtin"))
-	if err != nil {
-		return err
-	}
-	defer f.Close()
-
-	scanner := bufio.NewScanner(f)
-	for scanner.Scan() {
-		module := scanner.Text()
-		builtinModules[module] = true
+	for _, v := range modfiles {
+		if err := w.CopyFile(moddir + "/" + v); err != nil {
+			return err
+		}
 	}
 
-	if err := scanner.Err(); err != nil {
-		return err
+	prefix := ""
+	if mergedUsrSystem() {
+		prefix = "/usr"
 	}
 
-	for _, v := range modules {
-		if builtinModules[v] {
-			continue
-		}
+	copiedModules := make(map[string]bool)
 
-		modpath := path.Join(moddir, v)
+	size := len(modules)
+	for i := 0; i < size; i++  {
+		modname := modules[i]
 
-		if strings.HasSuffix(modpath, ".ko") {
-			if _, err := os.Stat(modpath); err != nil {
-				modpath += ".xz"
-			}
-			if _, err := os.Stat(modpath); err != nil {
-				return err
-			}
+		modpath := getModPath(modname)
+		if modpath == "" {
+			return errors.New("Modules path couldn't be determined")
+		}
+		if modpath == "(builtin)" || copiedModules[modname] {
+			continue
 		}
 
-		if err := w.CopyFile(modpath); err != nil {
+		if err := w.CopyFile(prefix + modpath); err != nil {
 			return err
 		}
+
+		copiedModules[modname] = true
+
+		// append dependent module of copied module
+		deplist := getModDepends(modname)
+		modules = append(modules, deplist...)
+		size += len(deplist)
```

Comment:
Thanks for fixing the dependency handling - now we properly handle the second+ level dependencies.

Not a big fan of looping over all array entries, while simultaneously adding extra ones at the end. So personally I would have made the contents of the loop a recursive standalone function. Some quick and untested example:

```
copyModule(modname...) {
    modpath := getModPath(modname)
    if modpath == "" {
        return errors.New("Modules path couldn't be determined")
    }
    if modpath == "(builtin)" || copiedModules[modname] {
        return
    }
    if err := w.CopyFile(prefix + modpath); err != nil {
        return err
    }

    copiedModules[modname] = true

    deplist := getModDepends(modname)
    for dep := range deplist {
        if err := copyModule(dep, ...); err != nil {
            return err
        }
    }
}
```

Reply:
done on latest push